pkg/globalcp/server.go

```diff
@@ -0,0 +1,130 @@
+package globalcp
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/Kong/kuma/pkg/core"
+)
+
+var (
+	globalCPLog = core.Log.WithName("global-cp")
+)
+
+type (
+	GlobalCP interface {
+		Start(<-chan struct{}) error
+	}
+
+	LocalCP struct {
+		URL    string `json:"url"`
+		Active bool   `json:"active"`
+	}
+
+	LocalCPMap map[string]*LocalCP
+
+	GlobalCPPoller struct {
+		sync.RWMutex
+		localCPMap LocalCPMap
+		server     *http.Server
+		newTicker  func() *time.Ticker
+	}
+)
+
+const (
+	tickInterval            = 5 * time.Second
+	globalCPStatisticsAddrt = ":5656"
+)
+
+func NewGlobalCPPoller(localCPList map[string]string) (GlobalCP, error) {
+	mux := http.NewServeMux()
+	poller := &GlobalCPPoller{
+		localCPMap: LocalCPMap{},
+		server: &http.Server{
+			Addr:    globalCPStatisticsAddrt,
+			Handler: mux,
+		},
+		newTicker: func() *time.Ticker {
+			return time.NewTicker(tickInterval)
+		},
+	}
+
+	for name, url := range localCPList {
+		poller.localCPMap[name] = &LocalCP{URL: url, Active: true}
+	}
+
+	mux.HandleFunc("/", poller.StatusHandler)
+	return poller, nil
+}
+
+func (g *GlobalCPPoller) Start(stop <-chan struct{}) error {
+	ticker := g.newTicker()
+	defer ticker.Stop()
+
+	// update the status before running the API
+	g.pollLocalCPs()
+
+	errChan := make(chan error)
+	go func() {
+		err := g.server.ListenAndServe()
+		if err != nil {
+			switch err {
+			case http.ErrServerClosed:
+				globalCPLog.Info("Shutting down server")
+			default:
+				globalCPLog.Error(err, "Could not start an HTTP Server")
+				errChan <- err
+			}
+		}
+	}()
+
+	globalCPLog.Info("starting the Global CP polling")
+	for {
+		select {
+		case <-ticker.C:
+			g.pollLocalCPs()
+		case <-stop:
+			globalCPLog.Info("Stopping down API Server")
+			return g.server.Shutdown(context.Background())
+		case err := <-errChan:
+			return err
+		}
+	}
+}
+
+func (g *GlobalCPPoller) pollLocalCPs() {
+	for name, localCP := range g.localCPMap {
```

Comment:
You probably need to RLock before iterating over the `g.localCPMap`

Reply:
Well, there is no dynamic update of the map, it is loaded from config at creating time.
On the other hand, all these global locks make the API endpoint being irresponsive.
Let me think if I can optimize it a bit.

Reply:
OK, global method lock aplied.